### ISSUE TITLE
Ulauncher v6

### ISSUE THREAD

**Issue by friday** [Ulauncher/Ulauncher#869]
### Description

This is a tracking issue for the v6 release that we have been working hard on. V6 is still in development and will not be released for quite a while due to containing major changes (and major improvements) and more changes planned.

The [6.0.0](https://github.com/Ulauncher/Ulauncher/milestone/7) milestone has a lot more information. It lists a lot of breaking changes, but they are mostly limited in scope such as moving niche CLI options, files for storing users settings and removing compatibility with very old distro releases.

This issue was created so that people interested in the progress can subscribe to the issue to be notified, since I will be closing issues that are fixed in the v6 branch and redirecting people here if they want a notification when we release or something similar.

### Feature isn't already requested

- [X] I have checked the existing feature requests before suggesting this

**Comment by joseph-zhong** [comment]
Huge shoutout and thanks to the team for working on this! Real excited for the v6 release; been using v5.x for only a few days and been a huge game-changer; particularly excited for the improvements in this release and also happy to beta-test -- is there suggested documentation for building/running from source?

**Comment by friday** [comment]
Thank you @joseph-zhong 

> Is there suggested documentation for building/running from source?

Yes: https://github.com/Ulauncher/Ulauncher/blob/v6/CONTRIBUTING.md

**Comment by friday** [comment]
You should be able to build a Debian package now with `./ul build-deb`, assuming you are on Debian and have the packages needed (see above for the general build dependencies and add `apt install debhelper dh-python rsync` for the debian package build.

**Comment by LinuxOnTheDesktop** [comment]
_Blacklisted dirs_ ('If you still feel like you need this setting in Ulauncher v6, let us know'). I think I do need this. For, there are directories that contain executables (or links to them) that do not work on my system or that I do not wish to use.

Also: could we have _dark mode in the configuration windows_, please?

**Comment by friday** [comment]
> _Blacklisted dirs_ ('If you still feel like you need this setting in Ulauncher v6, let us know'). I think I do need this. For, there are directories that contain executables (or links to them) that do not work on my system or that I do not wish to use.

Then why do you have them in the dedicated directory for desktop entries to use for launchers and menu apps if I may ask?

> Also: could we have _dark mode in the configuration windows_, please?

That's not planned, and is not easy to implement, so very unlikely for v6. Hopefully later at some point.

**Comment by LinuxOnTheDesktop** [comment]
> > Blacklisted dirs ('If you still feel like you need this setting in Ulauncher v6, let us know'). I think I do need this. For, there are directories that contain executables (or links to them) that do not work on my system or that I do not wish to use.
> 
> Then why do you have them in the dedicated directory for desktop entries to use for launchers and menu apps if I may ask?

I am unsure that I do. If (and to the extent that) I do, then I imagine that my OS (which is built on Ubuntu, which itself is built upon Debian) put them there. I implemented the blacklists so long ago that I forget why I did so. Here are the blacklisted dirs (I'd export them as text but I see no facility to do that):

![image](https://github.com/Ulauncher/Ulauncher/assets/47056543/7c69addd-b903-4d40-90dd-a4dc4046aa14)

Also: I imagine that Ulauncher implemented a blacklist for a reason . .

**Comment by lorenzleutgeb** [comment]
I might be a little late, but just checking: #712 and #713 should be resolved right. Asking because these issues are not mentioned. If so, I'd test the RC.

**Comment by friday** [comment]
> I might be a little late, but just checking: #712 and #713 should be resolved right. Asking because these issues are not mentioned. If so, I'd test the beta.

Yes. There's over 1000 commits, so I can't mention everything in the release notes, but they are closed as fixed so they are fixed.

**Comment by friday** [comment]
I just pushed a pre-release: https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-beta1. We still have work to do before the stable release, but this makes it easier for people to test without building themselves.

The plan was to also make that publish to [ulauncher-dev](https://launchpad.net/~agornostal/+archive/ubuntu/ulauncher-dev), for the people who opted for the experimental Ubuntu PPA. But apparently that part isn't working yet.

**Comment by friday** [comment]
https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-beta2 released and published to [ulauncher-dev](https://launchpad.net/~agornostal/+archive/ubuntu/ulauncher-dev)

**Comment by RedBearAK** [comment]
@friday 

Working nicely in a Xubuntu VM, but I'm hoping someone will start making some RPMs soon. My main systems are Fedora and Tumbleweed. The `tar.gz` is a bit too manual for my taste. 

Love the Ctrl+comma shortcut to open the preferences now. 

And the searching in the app "comment" or description appears to be working well also. Lots of good things happening.

**Comment by LinuxOnTheDesktop** [comment]
Version 6 sounds good! Where should bug reports for the beta of that version be posted, please? I add immediately here though the following two points (though neither are about bugs).

1) On Cinnamon, Ulauncher 5 itself could set the main Ulauncher shortcut. Ulauncher 6 cannot. But setting the shortcut via Cinnamon works.

2) Previously, for applications requiring pkexec, I had to create custom shortcuts else Ulauncher would not run those applications. But now Ulauncher runs those applications directly. Nice!

**Comment by friday** [comment]
> 1. On Cinnamon, Ulauncher 5 itself could set the main Ulauncher shortcut. Ulauncher 6 cannot. But setting the shortcut via Cinnamon works.

Yeah, that's because the old way relied on a library that relied on globally monitoring all keystrokes See #1256 and https://github.com/Ulauncher/Ulauncher/discussions/991#user-content-hotkey (old behavior). I added support for Cinnamon too, but had to revert it because it wasn't possible to reload Cinnamon to apply the settings: https://github.com/Ulauncher/Ulauncher/commit/2a227113dc68188df7496f568a3fdd0d5b5c4820

> 2. Previously, for applications requiring pkexec, I had to create custom shortcuts else Ulauncher would not run those applications. But now Ulauncher runs those applications directly. Nice!

:pray:

**Comment by friday** [comment]
[v6.0.0-beta3](https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-beta3) was released yesterday, but the [ulauncher-dev](https://launchpad.net/~agornostal/+archive/ubuntu/ulauncher-dev) ppa upload failed because of errors on their side. So the package wasn't pushed until some hours ago today.

**Comment by friday** [comment]
[v6.0.0-beta4](https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-beta4) just released with some fixes.

The scope of the v6 release still isn't fully set in stone, but we only have a few remaining planned **features/improvements**:
* https://github.com/Ulauncher/Ulauncher/issues/273#issuecomment-1038542547
* #931
* #1257 (this could be big, so maybe it will happen after v6, but it would be a huge benefit to get it shipped soon)

Most of the remaining tasks in the milestone are chores that relate to the release, extension site or documentation. While these can still take time, some can wait.

Also appreciate any help I can get with testing:
* Ensure the app shadow works as intended with a DE without a compositor
* Test more DEs and Wayland ([wlr-layer-shell](https://wayland.app/protocols/wlr-layer-shell-unstable-v1))
* According to one user Linux Mint seems to log apps launched by Ulauncher under "ulauncher" (we can likely not do something about this?)

**Comment by euh2** [comment]
> * According to one user Linux Mint seems to log apps launched by Ulauncher under "ulauncher" (we can likely not do something about this?)

I see this as well on Fedora. From my understanding it seems that Ulauncher is regarded as as parent process. There must be a way to dislodge the Child from the Parent. In the shell I would use `exec someApplication` to uncouple the application from my shell (or rather the binary running my shell). Maybe there is a way to do this in Python.

**Comment by RedBearAK** [comment]
@euh2 @friday 

> In the shell I would use exec someApplication to uncouple the application from my shell (or rather the binary running my shell). Maybe there is a way to do this in Python.

Actually from what I understand `exec` in the shell replaces the shell with the new process or shell, rather than running the command as a child or spawning another shell layer. I don't think it's quite the same as launching an independent process while the same shell keeps running. But I'm not an expert in how the shell works. 

For Python, GPT-4 suggests this may be the best way to launch a process and have it be independent of the Python process that launches it: 

```py
import subprocess
subprocess.Popen(['your_app'], start_new_session=True)
```

I've never had occasion to use this `start_new_session` argument, but it seems to be a valid argument and defaults to `False`. Seems to have been around since Python 3.2 (2011). 

> If start_new_session is true the setsid() system call will be made in the child process prior to the execution of the subprocess.

References to it found here: 

https://docs.python.org/3/library/subprocess.html

**Comment by friday** [comment]
We have just released https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-b5 which should fix the issue mentioned above (it also reintroduced #778 when running Ulauncher from a terminal) @euh2 @RedBearAK

This is the first release for a while because most of what's been done in this period has been refactoring and build improvements, NixOS support and other things which didn't affect the app so much.

There is however something wrong with launchpad again. The logs are saying that all the builds were successful and uploaded, but the PPA only has an upload for Ubuntu Noble (24.04), and I tested several times. It is probably time to find a better alternative to this frustrating platform.

**Comment by euh2** [comment]
> It is probably time to find a better alternative to this frustrating platform.

Suse might be an alternative: [openSuse](https://build.opensuse.org)

I'm curious about why I was tagged here. 🤔

**Comment by friday** [comment]
> I'm curious about why I was tagged here. 🤔

Because this release should address your feedback in https://github.com/Ulauncher/Ulauncher/issues/869#issuecomment-1894705268

> Suse might be an alternative: [openSuse](https://build.opensuse.org/)

Feels like a better alternative for sure, but extremely high threshold to learn how to use it. I tried it twice.

**Comment by euh2** [comment]
> should address your feedback

I forgot about that. Waiting eagerly for the stable v6. 🇺🇦

**Comment by RedBearAK** [comment]
@friday 

I continue to look forward to someone putting together an RPM installer. I spend most of my time on Fedora. I'm betting between openSUSE, OpenMandriva, Fedora, CentOS, RHEL and all the RHEL compatibles, the RPM-based distros are a decently large portion of the installed Linux user base. 

I tried just using the `setup.py` to get it built and installed, but the result seems to have a lot of Python exceptions, and I don't really have time to mess around with it to see why it's not working. I'm sure the DEB still works perfectly fine on Ubuntu/Debian systems. 

It also seems to be putting things in the "system" Python locations, which more and more distros are restricting in favor of wanting apps to use their own virtual environment to protect the system Python library. Seems like a good idea to sort that out before the final release. I had to go through that process with a project of mine about a year ago, and now it will install without any Python issues on about 6 different major distro types, using its own isolated `venv`.

**Comment by friday** [comment]
@RedBearAK Thanks for the testing and the feedback. These are not priorities for us currently though. Also, system packages _should_ use system dependencies unless it's a flatpak or something.

I guess we could look into supporting https://help.ubuntu.com/community/CheckInstall though, but those uses the user package paths, not the system ones.

**Comment by friday** [comment]
[v6.0.0-beta9](https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-beta9) was just released.

Since the last update here (the v6.0.0-beta4 release) we have done a lot of refactoring, fixes and added some features, but mainly the same bigger tasks remain. We have been trying to reduce the [milestone](https://github.com/Ulauncher/Ulauncher/milestone/7) by moving #1257 out of the V6 release so we can have a release candidate ready sooner.
* #832
* #1235 (need to confirm)
* https://github.com/Ulauncher/Ulauncher/issues/273#issuecomment-1038542547 (this is one of the big tasks that we planned for v6 originally, so we really should try to include it)
* #931 (might move this out of the v6 milestone too)

The other unfinished tasks in the milestone are chores that relate to the release, extension site or documentation.

We would still appreciate any help we can get with testing for DEs without a compositor, Wayland etc to ensure things are working there.

**Comment by varac** [comment]
Looking forward to v6! Since this issue now is nearly 3y old, please rescope and postpone all non-critical issues for after v6 release, to get this out of the door soon :rocket:, thanks !

**Comment by varac** [comment]
The reason I'd like to see a release rather sooner than later is that as an extension author I'd like to write extensions for API v3 when I start coding extension, so I don't need to update them in the future.
But there's not an option add add them to the registry at this point, which is a pity.

**Comment by Mark3xtrm** [comment]
When will it be possible to put the automatic light/dark theme change?
I'm on Linux Mint Cinnamon, and when I change the theme, it doesn't change the Ulauncher theme.
Could you please add this possibility?

**Comment by troycurtisjr** [comment]
> I continue to look forward to someone putting together an RPM installer. I spend most of my time on Fedora. I'm betting between openSUSE, OpenMandriva, Fedora, CentOS, RHEL and all the RHEL compatibles, the RPM-based distros are a decently large portion of the installed Linux user base.
> [...snip...]

@RedBearAK I just packaged up ulauncher 6.0.0-beta16 in [my copr](https://copr.fedorainfracloud.org/coprs/troycurtisjr/ulauncher/). I will also work on building out some automation or similar to help keep it up to date.

**Comment by RedBearAK** [comment]
@troycurtisjr 

I appreciate the effort to put that together. The COPR seems to work, as does the beta of Ulauncher 6 that ends up on my system. 

Unfortunately, despite Ulauncher having a much better UI than `krunner`, especially with a good theme applied, there are a lot of things in Plasma that will never show up until someone can make an extension that integrates `krunner` search results. So Ulauncher is not nearly as useful to me in Plasma as it used to be in GNOME and other DEs. 

Some of the useful things that won't show up in Ulauncher are: Individual settings modules (KCMs) from the Plasma System Settings app, and session control options (e.g., Log Out, Restart, Shut Down). Someone needs to make an extension that allows Ulauncher to access `krunner` results in addition to showing the usual list of applications. 

Anyway, at least I can run Ulauncher 6 on Fedora now, as an alternative, and without trying to compile it myself. Seems very fast and stable. 

I look forward to seeing updates come from the COPR at some point.

**Comment by choman** [comment]
Hopefully I am not repeating another ticket from somewhere else. I tried looking and found none

I am on both ubuntu 24.04 and 24.10 and the "save Hotkey" does not appear to work. I've seen this with earlier versions of v6, so I thought I would try again with beta16 and it still appears broke. 

I click the button and the popup appears. But when I click "save" nothing and the prescribed hotkey does not launch uLauncher

**Comment by friday** [comment]
@choman Please create a bug issue and state what Desktop Environment you are using in it.

**Comment by choman** [comment]
> @choman Please create a bug issue and state what Desktop Environment you are using in it.

Will do, thanks

**Comment by witt-bit** [comment]
Hello, everyone. I am a manjaro Linux user. To facilitate the installation of Ulauncher V6, I released the dev AUR package: https://aur.archlinux.org/packages/ulauncher-dev-bin, referenced from https://aur.archlinux.org/packages/ulauncher. After installation, everything works fine. Can someone help me find out if there is an error in the build and packaging command in the PKGBUILD file?

**Comment by witt-bit** [comment]
> > Hello, everyone. I am a manjaro Linux user. To facilitate the installation of Ulauncher V6, I released the dev AUR package: https://aur.archlinux.org/packages/ulauncher-dev-bin, referenced from https://aur.archlinux.org/packages/ulauncher. After installation, everything works fine. Can someone help me find out if there is an error in the build and packaging command in the PKGBUILD file?
> 
> Why? Just use https://aur.archlinux.org/packages/ulauncher-git. It's maintained by me and has all the right deps. It's what I use myself.为什么？只需使用https://aur.archlinux.org/packages/ulauncher-git 。它由我维护并拥有所有正确的部门。这是我自己用的。

OK，I will delete the `ulauncher-dev-bin` package. For other issues, please discuss them in the comment section of ulauncher-git.

**Comment by friday** [comment]
New beta just released: https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-beta18

Unfortunately, launchpad somehow keeps getting worse despite not really updating the UI or adding features. The last two releases it only published Ubuntu Noble and Jammy, while Oracular and Focal doesn't build. I just tried to force them to rebuild but that also didn't work, and I can't see find any error logs. It doesn't really make sense for specifically Oracular and Focal to fail twice in a row if it's some kind of dependency issue or something else on my end, since that's the newest and oldest distros and the in-between ones work (and also: our build server uses Focal, so if that didn't work then neither would our own builds).

If you are on one of the Distros that Launchpad doesn't seem to want to build for any more you should still be able to download and manually install the `.deb` file on the release page.
